### PR TITLE
virtualisation/qemu-vm: fix unescaped bash variable in shell script

### LINKED ISSUE #229729: `qemu-vm.nix` does not escape disk image path
### Describe the bug
The `startVM` script defined in [nixos/modules/virtualisation/qemu-vm.nix](https://github.com/NixOS/nixpkgs/blob/aad25290ee4655200f63206c2a507611a8e6338d/nixos/modules/virtualisation/qemu-vm.nix) contains a [part that creates a disk image](https://github.com/NixOS/nixpkgs/blob/aad25290ee4655200f63206c2a507611a8e6338d/nixos/modules/virtualisation/qemu-vm.nix#L117-L135) if it does not yet exist. The file name of the disk image is stored in the bash variable `NIX_DISK_IMAGE` which is [written unquoted into the bash script](https://github.com/NixOS/nixpkgs/blob/aad25290ee4655200f63206c2a507611a8e6338d/nixos/modules/virtualisation/qemu-vm.nix#L133) when creating the disk image. Thus, if `$NIX_DISK_IMAGE` contains a space, the disk image creation fails.

### Steps To Reproduce
The following is an easy way to reproduce the problem under macOS with a recent nixpkgs; not sure about a Linux example, but anything that triggers a VM creation via `qemu-vm.nix` in the current directory should do the trick.
1. Create and enter a folder with a space in its name: `mkdir ~/"foo bar" && cd ~/"foo bar"`
2. Start a VM via nix, e.g., `nix run nixpkgs#darwin.builder` (a recent builder VM for macOS).
3. (will ask for sudo password for key setup)
4. Disk image creation is triggered and fails with the following output:
    ```console
    Disk image do not exist, creating the virtualisation disk image...
    qemu-img: Invalid image size specified. You may use k, M, G, T, P or E suffixes for
    qemu-img: kilobytes, megabytes, gigabytes, terabytes, petabytes and exabytes.
    ```

### Expected behavior
The builder disk image for the VM should be created and the VM should be started. (Also seems like there is a minor typo in the error message: "Disk image ~~do~~ **does** not exist, …".)

### Notify maintainers
@RaitoBezarius 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 22.4.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.15.0`
 - channels(peter): `""`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by pitkling**
###### Description of changes

Fixes #229729 by properly escaping the shell variable `NIX_DISK_IMAGE` in the inlined shell script.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -130,7 +130,7 @@ let
           ${concatStringsSep " \\\n" ([ "-f qcow2" ]
           ++ optional (cfg.useBootLoader && cfg.useDefaultFilesystems) "-F qcow2 -b ${systemImage}/nixos.qcow2"
           ++ optional (!(cfg.useBootLoader && cfg.useDefaultFilesystems)) "-o size=${toString config.virtualisation.diskSize}M"
-          ++ [ "$NIX_DISK_IMAGE" ])}
+          ++ [ ''"$NIX_DISK_IMAGE"'' ])}
```
Comment (wegank) [inline nixos/modules/virtualisation/qemu-vm.nix:133]:
```suggestion
          ++ [ lib.escapeShellArg "$NIX_DISK_IMAGE" ])}
```

Comment (pitkling) [reply nixos/modules/virtualisation/qemu-vm.nix:133]: Ah, I didn't know that `lib` contains such helper functions.

Anyway, I don't think this works. I check the definition of `escapeShellArg`. The call `escapeShellArg "$NIX_DISK_IMAGE"` would result in the string `'$NIX_DISK_IMAGE'` (notice the **single** quotes) being written into the shell script. The shell variable would then no longer expand due to these single quotes.

I quickly checked this with a local nixpkgs (note: the call of `escapeShellArg "$NIX_DISK_IMAGE"` must be within parenthesis) and indeed, this results in a disk image file with name `$NIX_DISK_IMAGE` (unexpanded) created in the current directory.

But let me know if there is a better way than the two single quotes from my commit to escape the shell variable and maintain expansion, I'm not very familiar with nixpkgs's auxiliary functions.

**Review by wegank — APPROVED** [review]

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 2]
```diff
@@ -130,7 +130,7 @@ let
           ${concatStringsSep " \\\n" ([ "-f qcow2" ]
```
Comment (ajs124) [inline nixos/modules/virtualisation/qemu-vm.nix:130]: this should probably just be using `lib.escapeShellArgs`

Comment (pitkling) [reply nixos/modules/virtualisation/qemu-vm.nix:130]: Would that not again lead to the `$NIX_DISK_IMAGE` shell variable being put in between single quotes and prevent shell expansion in the resulting script, similar to [this previous suggestion](https://github.com/NixOS/nixpkgs/pull/229845/commits/996db6d059cd8e6f3b5815a33816bf8ea16f42b5#r1184961872)?

**Pull request merged**